### PR TITLE
build(releases): automate next version increment

### PR DESCRIPTION
**Pull request by brainbicycle**
This PR resolves [PHIRE-875] 

### Description

Attempts to automate the step to create the next version of the app.

As part of the nightly beta job check if there is an app version 'Ready for Sale' but not an app version in 'Prepare for Submission'. This should be the state after an app is approved by apple for distribution but we have not yet created the next version for development. If so create a new version. 

#### Note on testing!

Because this depends on our app not having an editable version ready I could only test the individual parts of this. I tried to be thorough but the last bit, actually creating the next version via fastlane, I think we just have to see the next time we release if it works.

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- automate next version creation - brian 

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

[PHIRE-875]: https://artsyproduct.atlassian.net/browse/PHIRE-875?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: fastlane/Fastfile [hunk 1]
```diff
@@ -174,21 +176,38 @@ lane :set_git_properties_ios do
...
+  live_version = app.get_live_app_store_version(platform: "iOS")
+
+  edit_version_ready = edit_version && edit_version.app_store_state == 'PREPARE_FOR_SUBMISSION'
+  live_version_ready = live_version && live_version.app_store_state == 'READY_FOR_SALE'
+  if live_version_ready and !edit_version_ready then
+    UI.info("There is a live version that is ready for sale, but there is no editable version")
+    UI.info("Creating new version #{next_version}")
+    next_version = increment_version_number(live_version.version_string)
+    update_version_string(version: next_version)
+    prepare_version_update_pr(commit_message: "chore: prepare for next release")
```
Comment (brainbicycle) [inline fastlane/Fastfile:196]: Might be nice to send a slack message with a link to the pr when we do this

File: fastlane/Fastfile [hunk 2]
```diff
@@ -174,21 +176,45 @@ lane :set_git_properties_ios do
...
+  live_version = app.get_live_app_store_version(platform: "iOS")
+
+  edit_version_ready = edit_version && edit_version.app_store_state == 'PREPARE_FOR_SUBMISSION'
+  live_version_ready = live_version && live_version.app_store_state == 'READY_FOR_SALE'
+  if live_version_ready and edit_version_ready then
+    UI.message("There is a live version that is ready for sale, but there is no editable version")
+    next_version = increment_version_number(live_version.version_string)
+    UI.message("Creating new version #{next_version}")
+    update_version_string(version: next_version)
+    app.create_app_store_version(platform: "iOS", version_string: next_version)
```
Comment (brainbicycle) [inline fastlane/Fastfile:196]: made up method courtesy of chatGPT 😭

File: fastlane/Fastfile [hunk 3]
```diff
@@ -174,21 +176,45 @@ lane :set_git_properties_ios do
...
+
+  app = Spaceship::ConnectAPI::App.find(BUNDLE_ID)
+
+  edit_version = app.get_edit_app_store_version(platform: "iOS")
+  live_version = app.get_live_app_store_version(platform: "iOS")
+
+  edit_version_ready = edit_version && edit_version.app_store_state == 'PREPARE_FOR_SUBMISSION'
+  live_version_ready = live_version && live_version.app_store_state == 'READY_FOR_SALE'
+  if live_version_ready and !edit_version_ready then
+    UI.message("There is a live version that is ready for sale, but there is no editable version")
```
Comment (damassi) [inline fastlane/Fastfile:192]: Does it make sense to use "sale" here?

File: fastlane/Fastfile [hunk 4]
```diff
@@ -174,21 +176,45 @@ lane :set_git_properties_ios do
...
+  MSG
+    slack(message: message, default_payloads: [], link_names: true)
+  else
+    if (edit_version_ready) then
+      UI.success("There is an editable version ready for betas.")
+    else
+      message = <<~MSG
+      :x: :iphone:
+      New app version could not be created.
+      Create a new version in app store connect to keep shipping betas.
```
Comment (damassi) [inline fastlane/Fastfile:212]: Would it be helpful to include the reason why the version couldn't be created in the output message?

**Review by damassi — DISMISSED** [review]
Couple minor qs, tho nothing blocking. This is a really nice addition 👌

File: fastlane/Fastfile (hunk 3 above)
Comment (brainbicycle) [reply fastlane/Fastfile:192]: that is apple's terminology but maybe can just say "There is a live version but there is no editable version"

File: fastlane/Fastfile (hunk 4 above)
Comment (brainbicycle) [reply fastlane/Fastfile:212]: I can put something generic but not 100% on all the reasons that might happen, basically the live version or edit version is in a state we don't expect, maybe most likely is an update got rejected.

Comment (damassi) [reply fastlane/Fastfile:212]: Yah, just listing a few possibilities might make things a bit less opaque, even if its not 100% correct 👍

Comment (brainbicycle) [reply fastlane/Fastfile]: updated, let me know if you want anything else!

File: fastlane/Fastfile [hunk 5]
```diff
@@ -453,20 +481,6 @@ def set_codepush_release_values(deployment_name, platform)
...
-  ios_build_number = latest_testflight_build_number(
-    api_key: api_key
-  )
-  android_build_number = google_play_track_version_codes(
-      track: 'alpha'
-  ).first
-  puts "Parent latest version: #{parent_version_name}"
-  puts "iOS latest version: #{ios_build_number}"
-  puts "Android latest version: #{android_build_number}"
-end
```
Comment (brainbicycle) [inline fastlane/Fastfile:468]: moved to utils file

File: fastlane/Fastfile (hunk 4 above)
Comment (damassi) [reply fastlane/Fastfile]: Perfect 👍

**Review by damassi — DISMISSED** [review]

File: fastlane/Fastfile [hunk 6]
```diff
@@ -174,21 +176,47 @@ lane :set_git_properties_ios do
...
+      :x: :iphone:
+      New app version could not be created.
+      Possible reasons: The app was rejected, or the app is not in a state to be edited.
+      Check app store connect for specific details.
+      Once resolved, create a new version in app store connect to keep shipping betas.
+      https://appstoreconnect.apple.com/apps/703796080
+    MSG
+      slack(message: message, default_payloads: [], link_names: true)
+    end
+  end
```
Comment (brainbicycle) [inline fastlane/Fastfile:219]: this method is the meat of it:
```
if live app and no editable app (the next version to go out):
   - create a new version in app store connect
   - update the app.json with the new version
   - open a pr to update in main
   - send a slack message to remind us to merge the pr
else:
   - if (edit_version)
      -  we are good to go 
   - else
      - send a slack message that we need an edit version and it couldn't be created for some reason
```

File: fastlane/Fastfile [hunk 7]
```diff
@@ -174,21 +176,47 @@ lane :set_git_properties_ios do
...
+    if (edit_version_ready) then
+      UI.success("There is an editable version ready for betas.")
+    else
+      message = <<~MSG
+      :x: :iphone:
+      New app version could not be created.
+      Possible reasons: The app was rejected, or the app is not in a state to be edited.
+      Check app store connect for specific details.
+      Once resolved, create a new version in app store connect to keep shipping betas.
+      https://appstoreconnect.apple.com/apps/703796080
```
Comment (MrSltun) [inline fastlane/Fastfile:215]: **Suggestion:** This link sends me to a page with an error 👀 
should we change it to `https://appstoreconnect.apple.com/apps/703796080/appstore`

File: fastlane/utility_fastlane.rb [hunk 8]
```diff
@@ -74,10 +74,67 @@
...
+  sh "git push origin HEAD:refs/heads/#{version_change_branch}"
+
+  pr_url = create_pull_request(
+    api_token: ENV["CHANGELOG_GITHUB_TOKEN_KEY"],
+    repo: "artsy/eigen",
+    title: commit_message,
+    head: version_change_branch,
+    assignees: ["brainbicycle", "gkartalis"],
+    base: "main",
+    body: "This PR updates the app version to prepare for next release."
```
Comment (MrSltun) [inline fastlane/utility_fastlane.rb:118]: **Suggestion:** maybe we could add `#nochagelog` to the body so it shows up in the changelog script as no change log?

File: fastlane/Fastfile [hunk 9]
```diff
@@ -174,21 +176,47 @@ lane :set_git_properties_ios do
...
+    next_version = increment_version_number(live_version.version_string)
+    UI.message("Creating new version #{next_version}")
+    update_version_string(version: next_version)
+    create_next_app_version(next_version_code:  next_version)
+    pr_url = prepare_version_update_pr(commit_message: "chore: prepare for next release")
+    message = <<~MSG
+    :x: :iphone:
+    It's time to update the app version number!
+    Merge this pr to keep shipping betas:
+    #{pr_url}
```
Comment (MrSltun) [inline fastlane/Fastfile:202]: **Praise:** 👏

**Review by MrSltun — APPROVED** [review]
Great addition Brian!! 🌟🌟🌟
I just left a few minor non-blocking suggestions

File: fastlane/Fastfile (hunk 7 above)
Comment (brainbicycle) [reply fastlane/Fastfile:215]: good catch!

File: .circleci/config.yml [hunk 10]
```diff
@@ -176,13 +176,22 @@ jobs:
...
-    executor:
-      name: node/default
-      tag: "20.9.0"
-    resource_class: small
+    environment:
+      BUNDLE_PATH: .vendor
+
+    macos:
+      xcode: 15.1
+      resource_class: macos.m1.medium.gen1
```
Comment (brainbicycle) [inline .circleci/config.yml:184]: since we need to use fastlane here had to update the executor

File: fastlane/Fastfile [hunk 11]
```diff
@@ -174,21 +176,47 @@ lane :set_git_properties_ios do
...
+  live_version_ready = live_version && live_version.app_store_state == 'READY_FOR_SALE'
+  if live_version_ready and !edit_version_ready then
+    UI.message("There is a live app version but there is no editable version")
+    next_version = increment_version_number(live_version.version_string)
+    UI.message("Creating new version #{next_version}")
+    update_version_string(version: next_version)
+    create_next_app_version(next_version_code:  next_version)
+    pr_url = prepare_version_update_pr(commit_message: "chore: prepare for next release")
+    message = <<~MSG
+    :x: :iphone:
```
Comment (brainbicycle) [inline fastlane/Fastfile:199]: ❌ implies error, pick a better emoji 🚂

**Pull request merged**